File: development/NET/samples/server/AddressSpace/CS/MyDaAddressSpaceElement.cs
using System;
using System.Collections;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace AddressSpace
{
	public class MyDaAddressSpaceElement : DaAddressSpaceElement
	{

		public MyDaAddressSpaceElement(
			string anItemID,
			string aName,
			uint anUserData,
			uint anObjectHandle,
			uint aParentHandle) :
			base(anItemID, aName, anUserData,
			anObjectHandle, aParentHandle)
		{
			this.AddedToAddressSpace += new AddedToAddressSpace(HandleAddedToAddressSpace);
			this.RemovedFromAddressSpace += new RemovedFromAddressSpace(HandleRemovedFromAddressSpace);
		}   //	end ctr

		public MyDaAddressSpaceElement()
		{
			this.AddedToAddressSpace += new AddedToAddressSpace(HandleAddedToAddressSpace);
			this.RemovedFromAddressSpace += new RemovedFromAddressSpace(HandleRemovedFromAddressSpace);
		}   //	end ctr

		public override int QueryProperties(out ArrayList aPropertyList)
		{

			aPropertyList = new ArrayList();

			DaProperty prop = new DaProperty();
			prop.Id = 5555;
			prop.AccessRights = EnumAccessRights.READABLE;
			prop.Datatype = typeof(UInt16);
			prop.Name = "type";
			prop.Description = "Object Type";
			if (ItemId == null)
			{
				prop.ItemId = "objectType";
			}
			else
			{
				prop.ItemId = ItemId + Application.Instance.PropertyDelimiter + "objectType";
			}   //	end if ... else

			DaProperty prop2 = new DaProperty();
			prop2.Id = 5556;
			prop2.AccessRights = EnumAccessRights.READABLE;
			prop2.Datatype = typeof(UInt16);
			prop2.Name = "type2";
			prop2.Description = "Object Type2";
			if (ItemId == null)
			{
				prop2.ItemId = "objectType2";
			}
			else
			{
				prop2.ItemId = ItemId + Application.Instance.PropertyDelimiter + "objectType2";
			}   //	end if ... else

			aPropertyList.Add(prop);
			aPropertyList.Add(prop2);

			return (int)EnumResultCode.S_OK;

		}   //	end QueryProperties

		public void HandleAddedToAddressSpace()
		{

			if (ItemId == "stringBased")
			{
				ObjectType = 20;
			}
			else if ((ItemId == "stringBased.N1") || (ItemId == "stringBased.N2"))
			{
				ObjectType = 21;
			}
			else if (ItemId == "stringBased.T1")
			{
				ObjectType = 22;
			}
			else if ((ItemId == "stringBased.N1.T1") ||
				(ItemId == "stringBased.N2.T1") ||
				(ItemId == "stringBased.N2.T2"))
			{
				ObjectType = 23;
			}
			else if (ItemId == "syntaxBased")
			{
				ObjectType = 40;
			}
			else if (ItemId.StartsWith("syntaxBased"))
			{
				Int32 index;
				index = ItemId.LastIndexOf(".T");
				if (index > 0)
				{
					ObjectType = 42; // T
				}
				else
				{
					ObjectType = 41; // N
				}   //	end if ... else
			}   //	end if ... else



			if (this.IoMode != EnumIoMode.NONE)
			{
				ValueQT initialValue = new ValueQT(0, EnumQuality.GOOD, DateTime.Now);
				ValueChanged(initialValue);
			}

			OpcServer.AddressSpaceChanged.Set();

		}   //	end HandleAddedToAddressSpace

		public void HandleRemovedFromAddressSpace()
		{
			OpcServer.AddressSpaceChanged.Set();
		}   //	end RemovedFromAddressSpace

		private byte m_objectType;

		public byte ObjectType
		{
			get { return m_objectType; }
			set { m_objectType = value; }
		}   //	end ObjectType

	}   //	end class MyDaAddressSpaceElement

}   //	end namespace

File: development/NET/samples/server/AddressSpace/VB/MyDaAddressSpaceElement.vb
Imports System
Imports System.Collections
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace AddressSpace
	Public Class MyDaAddressSpaceElement
		Inherits DaAddressSpaceElement

		Public Sub New(ByVal anItemID As String, ByVal aName As String, ByVal anUserData As UInt32, ByVal anObjectHandle As UInt32, ByVal aParentHandle As UInt32)
			MyBase.New(anItemID, aName, anUserData, anObjectHandle, aParentHandle)
			AddHandler Me.AddedToAddressSpace, AddressOf HandleAddedToAddressSpace
			AddHandler Me.RemovedFromAddressSpace, AddressOf HandleRemovedFromAddressSpace
		End Sub

		Public Sub New()
			AddHandler Me.AddedToAddressSpace, AddressOf HandleAddedToAddressSpace
			AddHandler Me.RemovedFromAddressSpace, AddressOf HandleRemovedFromAddressSpace
		End Sub

		Public Overloads Overrides Function QueryProperties(ByRef aPropertyList As ArrayList) As Integer

			aPropertyList = New ArrayList()

			Dim prop As New DaProperty()
			prop.Id = 5555
			prop.AccessRights = EnumAccessRights.READABLE
			prop.Datatype = GetType(UInt16)
			prop.Name = "type"
			prop.Description = "Object Type"
			If ItemId Is Nothing Then
				prop.ItemId = "objectType"
			Else
				prop.ItemId = ItemId + Application.Instance.PropertyDelimiter + "objectType"
			End If

			Dim prop2 As New DaProperty()
			prop2.Id = 5556
			prop2.AccessRights = EnumAccessRights.READABLE
			prop2.Datatype = GetType(UInt16)
			prop2.Name = "type2"
			prop2.Description = "Object Type2"
			If ItemId Is Nothing Then
				prop2.ItemId = "objectType2"
			Else
				prop2.ItemId = ItemId + Application.Instance.PropertyDelimiter + "objectType2"
			End If

			aPropertyList.Add(prop)
			aPropertyList.Add(prop2)

			Return EnumResultCode.S_OK

		End Function

		Public Sub HandleAddedToAddressSpace()

			If ItemId = "stringBased" Then
				ObjectType = 20
			ElseIf (ItemId = "stringBased.N1") OrElse (ItemId = "stringBased.N2") Then
				ObjectType = 21
			ElseIf ItemId = "stringBased.T1" Then
				ObjectType = 22
			ElseIf (ItemId = "stringBased.N1.T1") OrElse (ItemId = "stringBased.N2.T1") OrElse (ItemId = "stringBased.N2.T2") Then
				ObjectType = 23
			ElseIf ItemId = "syntaxBased" Then
				ObjectType = 40
			ElseIf ItemId.StartsWith("syntaxBased") Then
				Dim index As Int32
				index = ItemId.LastIndexOf(".T")
				If index > 0 Then
					ObjectType = 42
				Else
					ObjectType = 41
				End If
			End If



			If Me.IoMode <> EnumIoMode.NONE Then
				Dim qualityGood As EnumQuality = CType([Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD), EnumQuality)
				Dim initialValue As New ValueQT(0, qualityGood, DateTime.Now)
				ValueChanged(initialValue)
			End If

			OpcServer.AddressSpaceChanged.[Set]()

		End Sub

		Public Sub HandleRemovedFromAddressSpace()
			OpcServer.AddressSpaceChanged.[Set]()
		End Sub

		Private m_objectType As Byte

		Public Property ObjectType() As Byte
			Get
				Return m_objectType
			End Get
			Set(ByVal value As Byte)
				m_objectType = value
			End Set
		End Property

	End Class

End Namespace

